File: leetcode_2022/udemy_leetcode/graph_bfs_dfs.js
//adjcency list
const adjList = [
    [1,3],
    [0],
    [3,8],
    [0,4,5,2],
    [3,6],
    [3],
    [4,7],
    [6],
    [2]
]

const traversalBFS = (graph) => {
    const queue = [0]; //initial value to start
    const values = [];
    const seen = {};

    while (queue.length){
        const vertex = queue.shift();
        values.push(vertex);
        seen[vertex] = true;
        
        const connections = graph[vertex];
        
        for(let i = 0; i < connections.length; i++){
            const connection = connections[i];
            if(!seen[connection]){
                queue.push(connection)
            }
        }
    }
    return values;
}


const traversalDFS = (vertex, graph, values, seen) => {
    values.push(vertex);
    seen[vertex] = true;
    
    const connections = graph[vertex];
    
    for(let i = 0; i < connections.length; i++){
        const connection = connections[i];
        if(!seen[connection]){
            traversalDFS(connection, graph, values, seen)
        }
    }
}


console.log(traversalBFS(adjList))

//dfs is recursive, just like for many other data structure..
const values = []
const seen = {}
traversalDFS(0, adjList, values, seen)
console.log(values)

File: leetcode_2022/udemy_leetcode/graph_bfs_dfs.py
adjList = [
    [1,3],
    [0],
    [3,8],
    [0,4,5,2],
    [3,6],
    [3],
    [4,7],
    [6],
    [2]
]

def traversalBFS(graph):
    queue = [0]
    values = []
    seen = {}
    
    while len(queue):
        vertex = queue.pop(0)
        values.append(vertex)
        seen[vertex] = True
        
        connections = graph[vertex]
        
        for connection in connections:
            if connection not in seen:
                queue.append(connection)
                
    return values

print(traversalBFS(adjList))


def traversalDFS(vertex, graph, values, seen):
    values.append(vertex)
    seen[vertex] = True
    
    connections = graph[vertex]
    
    for connection in connections:
        if connection not in seen:
            traversalDFS(connection, graph, values, seen)
            

values = []
seen = {}

traversalDFS(0, adjList, values, seen)
print(values)